### PR TITLE
ODBC Driver: SQLColAttribute

### PR DESCRIPTION
**Pull request by tiagokepe**
Hey all,

This PR adds support for the colattribute-test, and it also fixes issue #2015.

Also, we support row-wise fetching that is defined by setting the statement attribute _SQL_ATTR_ROW_BIND_TYPE_. Details are disclosed in the test case: **row-wise-fetching-test**.

### CODE REVIEW

File: tools/odbc/statement_functions.cpp [hunk 1]
```diff
@@ -97,11 +97,14 @@ static void LogInvalidCast(const LogicalType &from_type, const LogicalType &to_t
 
 template <class T>
 static SQLRETURN GetInternalValue(duckdb::OdbcHandleStmt *stmt, const duckdb::Value &val, const LogicalType &type,
-                                  SQLLEN buffer_length, SQLPOINTER target_value_ptr) {
+                                  SQLLEN buffer_length, SQLPOINTER target_value_ptr, SQLLEN *str_len_or_ind_ptr) {
 	D_ASSERT(((size_t)buffer_length) >= sizeof(T));
 	try {
 		auto casted_value = val.CastAs(type).GetValue<T>();
 		Store<T>(casted_value, (duckdb::data_ptr_t)target_value_ptr);
+		if (str_len_or_ind_ptr) {
```
Comment (Mytherin) [inline tools/odbc/statement_functions.cpp:105]: Should this be here still?

File: tools/odbc/statement.cpp [hunk 2]
```diff
@@ -144,7 +161,102 @@ SQLRETURN SQLColumns(SQLHSTMT statement_handle, SQLCHAR *catalog_name, SQLSMALLI
...
+			auto out_len = snprintf((char *)character_attribute_ptr, buffer_length, "%s", col_name.c_str());
+			if (string_length_ptr) {
+				*string_length_ptr = out_len;
+			}
+
+			return SQL_SUCCESS;
+		}
+		case SQL_DESC_OCTET_LENGTH:
+			// 0 DuckDB doesn't provide octet length
+			*numeric_attribute_ptr = 0;
```
Comment (Mytherin) [inline tools/odbc/statement.cpp:190]: Should we check if `numeric_attribute_ptr` is nullptr?

File: tools/odbc/CMakeLists.txt [hunk 3]
```diff
@@ -10,6 +10,9 @@ include_directories(${ODBC_INCLUDE_DIRS} include)
 
 add_library(
   duckdb_odbc SHARED
+  odbc_fetch.cpp # should be first because of foward declaration, i.e., circular
```
Comment (Mytherin) [inline tools/odbc/CMakeLists.txt:13]: This seems odd to me. Are you sure this is the case? Unless something funky is happening in the headers the order of `.cpp` files should not matter.

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great. Some comments:

File: tools/odbc/statement_functions.cpp (hunk 1 above)
Comment (tiagokepe) [reply tools/odbc/statement_functions.cpp:105]: I think so. We need to check if a pointer to the length of the data was provided, then fill it with the number of bytes written: 

"_For all other data types, this is the length of the data after conversion; that is, it is the size of the type to which the data was converted._" 

 https://docs.microsoft.com/en-us/sql/odbc/reference/syntax/sqlgetdata-function?view=sql-server-ver15#:~:text=For%20all%20other%20data%20types%2C%20this%20is%20the%20length%20of%20the%20data%20after%20conversion%3B%20that%20is%2C%20it%20is%20the%20size%20of%20the%20type%20to%20which%20the%20data%20was%20converted.

File: tools/odbc/statement.cpp (hunk 2 above)
Comment (tiagokepe) [reply tools/odbc/statement.cpp:190]: Good catch, I will add that check.

File: tools/odbc/CMakeLists.txt (hunk 3 above)
Comment (tiagokepe) [reply tools/odbc/CMakeLists.txt:13]: I added just as a precaution, but I did a test and the order of .cpp doesn't matter. I will remove that comment.
Thanks for the tip.

File: tools/odbc/statement_functions.cpp (hunk 1 above)
Comment (Mytherin) [reply tools/odbc/statement_functions.cpp:105]: Apologies, this wasn't entirely clear. I meant the fact that the line after this was commented out:

```cpp
if (str_len_or_ind_ptr) {
  // *str_len_or_ind_ptr = sizeof(T);
}
```

Comment (tiagokepe) [reply tools/odbc/statement_functions.cpp:105]: I see, I forgot to remove the comment symbols (//). I changed in my branch, but I don't commit yet.

**Comment by Mytherin** [comment]
Thanks, looks good!

**Pull request merged**